File: src/bible_utils.py
"""
src/bible_utils.py
------------------
Shared utilities for bible file parsing and deduplication.

The Korean name/term is the canonical unique identifier for every bible entry.
This module provides the single source of truth for heading key extraction,
used by all modules that read or write bible files.
"""

import re


def heading_key(raw_heading: str) -> str:
    """
    Derive a canonical deduplication key from a raw ## heading string
    (everything after the leading "## ").

    Key selection priority:
    1. If the heading contains a parenthetical with Korean characters,
       use the Korean string as the key (lowercased, stripped).
       This makes the Korean name the single source of truth regardless
       of how the English romanisation is spelled.
    2. Otherwise, normalise the English text:
       - Strip everything after " — " (drops "— English" label suffixes)
       - Drop any remaining parentheticals
       - Lowercase and strip whitespace

    Examples
    --------
    "Hee-yeon Lee (이희연) — English"  → "이희연"
    "LOAN (로안) — English"            → "로안"
    "Ro-an (로안) — English"           → "로안"   ← same key as above
    "SM Entertainment — English"       → "sm entertainment"
    "Debut — English"                  → "debut"
    """
    korean_match = re.search(
        r"\(([^)]*[\uAC00-\uD7A3\u1100-\u11FF\u3130-\u318F][^)]*)\)",
        raw_heading
    )
    if korean_match:
        return korean_match.group(1).strip().lower()

    text = raw_heading
    text = re.split(r"\s+[—–-]\s+", text)[0]
    text = re.sub(r"\(.*?\)", "", text)
    return text.strip().lower()


def extract_heading_keys(text: str) -> set[str]:
    """
    Return the set of canonical deduplication keys for all ## headings
    found in a markdown string.

    Parameters
    ----------
    text : str
        Full contents of a bible file or a single entry block.

    Returns
    -------
    set[str]
        One key per ## heading found, derived via heading_key().
    """
    keys = set()
    for match in re.finditer(r"^##\s+(.+)$", text, re.MULTILINE):
        keys.add(heading_key(match.group(1)))
    return keys


def extract_headings_with_keys(text: str) -> list[tuple[str, str]]:
    """
    Return a list of (raw_heading, key) pairs for all ## headings found
    in a markdown string. Used for diagnostic/audit purposes.

    Parameters
    ----------
    text : str
        Full contents of a bible file.

    Returns
    -------
    list[tuple[str, str]]
        Each item is (raw heading text, canonical key).
    """
    results = []
    for match in re.finditer(r"^##\s+(.+)$", text, re.MULTILINE):
        raw = match.group(1)
        results.append((raw, heading_key(raw)))
    return results

File: app/services/pipeline/bible_utils.rb
# ---------------------------------------------------------------------------
# Pipeline::BibleUtils
#
# Ruby port of src/bible_utils.py — the single source of truth for
# deriving a canonical dedup key from a bible-entry "## heading" line.
# The Korean name/term is the canonical unique identifier for every bible
# entry; this module is what every reader/writer of bible files (R6's
# PrereadBibleWriter now, R5's BibleReviewWriter later) keys dedup on.
# ---------------------------------------------------------------------------
module Pipeline
  module BibleUtils
    KOREAN_PARENTHETICAL = /\(([^)]*[가-힣ᄀ-ᇿ㄰-㆏][^)]*)\)/
    HEADING_LINE = /^##\s+(.+)$/

    # Derive a canonical dedup key from a raw "## heading" string (everything
    # after the leading "## "). Priority: a parenthetical containing Korean
    # characters wins (lowercased, stripped) — this makes the Korean name the
    # single source of truth regardless of English romanisation. Otherwise,
    # normalise the English text: drop everything after " — " / " - " / " – ",
    # strip any remaining parentheticals, lowercase and strip.
    def self.heading_key(raw_heading)
      korean_match = KOREAN_PARENTHETICAL.match(raw_heading)
      return normalize_korean(korean_match[1]) if korean_match

      text = raw_heading.split(/\s+[—–-]\s+/, 2).first || raw_heading
      text = text.gsub(/\(.*?\)/, "")
      normalize_korean(text)
    end

    # Normalise a Korean (or mixed Korean/Latin) string for stable identity
    # comparison — a dedup key, a preread_dismissed_keys entry, a bible-entry
    # lookup key. The LLM's exact rendering of the same underlying term can
    # drift run to run (stray whitespace, full-width vs half-width chars,
    # Latin-script casing) even though the term itself hasn't changed; every
    # caller that treats Korean text as an identity — not just as display
    # text — should normalise through this one method so "the same term"
    # reliably produces "the same key". Unicode-normalise (NFKC), collapse
    # internal whitespace, strip, and downcase (a no-op on Hangul, but keeps
    # any Latin-script terms case-insensitive too).
    def self.normalize_korean(str)
      str.to_s.unicode_normalize(:nfkc).gsub(/\s+/, " ").strip.downcase
    end

    # Return the set of canonical dedup keys for all ## headings in a
    # markdown string (a full bible file, or a single entry block).
    def self.extract_heading_keys(text)
      Set.new(text.scan(HEADING_LINE).map { |match| heading_key(match.first) })
    end
  end
end
